sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ConcurrentConsumer.java

```diff
@@ -0,0 +1,231 @@
+          consumer.assign(nextAssignment);
+        }
+
+        positionsView.forEach(
+            (tp, o) -> {
+              if (o == Long.MIN_VALUE) {
+                consumer.pause(Collections.singleton(tp));
+              } else if (!prevAssignment.contains(tp)) {
+                consumer.seek(tp, o);
+              }
+            });
+
+        if (consumer.paused().size() != nextAssignment.size()) {
+          pollResult = consumer.poll(pollDuration.toMillis());
+        }
+
+        nextAssignment.forEach(tp -> positions.put(tp, consumer.position(tp)));
+        return false;
+      } catch (WakeupException e) {
+        if (!this.isTerminated()) {
+          LOG.error("Unexpected wakeup while running", e);
+        }
+      } catch (Exception e) {
+        LOG.error("Exception while reading from Kafka", e);
+      }
+      return true;
+    }
+  }
+
+  ConcurrentConsumer(final Consumer<K, V> consumer, final Duration pollDuration) {
+    this.phaser = new ConsumerPhaser();
+    this.consumer = consumer;
+    this.pollDuration = pollDuration;
+    this.times = AtomicLongMap.create();
+    this.positions = AtomicLongMap.create();
+    this.recordsLagMax =
+        Suppliers.memoize(
+            () ->
+                this.consumer.metrics().values().stream()
+                    .filter(
+                        m ->
+                            "consumer-fetch-manager-metrics".equals(m.metricName().group())
+                                && "records-lag-max".equals(m.metricName().name())
+                                && !m.metricName().tags().containsKey("topic")
+                                && !m.metricName().tags().containsKey("partition"))
+                    .findAny()
+                    .get());
+    this.partitionRecordsLag = new ConcurrentHashMap<>();
+    this.pollResult = ConsumerRecords.empty();
+    this.offsetsForTimesResult = Collections.emptyMap();
+  }
+
+  @Override
+  public void close() {
+    this.phaser.forceTermination();
+    try {
+      this.consumer.wakeup();
+      this.consumer.close();
+    } catch (Exception e) {
+      LOG.error("Exception while closing Kafka consumer", e);
+    }
+    this.times.clear();
+    this.positions.clear();
+    this.pollResult = ConsumerRecords.empty();
+    this.offsetsForTimesResult = Collections.emptyMap();
+  }
+
+  boolean isClosed() {
+    return this.phaser.isTerminated();
+  }
+
+  Map<MetricName, ? extends Metric> metrics() {
+    return this.consumer.metrics();
+  }
+
+  long currentLagOrMaxLag(final TopicPartition topicPartition) {
+    final Supplier<Metric> metric =
+        this.partitionRecordsLag.getOrDefault(topicPartition, this.recordsLagMax);
+    try {
+      return ((Number) metric.get().metricValue()).longValue();
+    } catch (Exception e) {
+      return 0;
+    }
+  }
+
+  long position(final TopicPartition topicPartition) {
+    return this.positions.get(topicPartition) & Long.MAX_VALUE;
```

Comment:
comment on the purpose of &, looks like to protect against exposing negative value?

Reply:
Correct, but I don't think we should keep this style of pause/resume bookkeeping since it's the least intuitive way to express it for readers and maintainers. It's unlikely that the representation of partition offsets will ever switch from signed to unsigned integers in a future version of Kafka, but not introducing a dependency on that assumption is also one less thing to worry about in the future.